src/test/java/com/puppycrawl/tools/checkstyle/internal/ImmutabilityTest.java

```diff
@@ -0,0 +1,467 @@
+         */
+        private static boolean isModuleProperty(JavaField javaField) {
+            boolean result = false;
+            final JavaClass containingClass = javaField.getOwner();
+            final ModuleDetails moduleDetails = MODULE_DETAILS_MAP.get(
+                containingClass.getFullName());
+            if (moduleDetails != null) {
+                final List<ModulePropertyDetails> properties = moduleDetails.getProperties();
+                result = properties.stream()
+                    .map(ModulePropertyDetails::getName)
+                    .anyMatch(moduleName -> moduleName.equals(javaField.getName()));
+            }
+            return result;
+        }
+
+        @Override
+        public boolean apply(JavaField input) {
+            return isModuleProperty(input);
+        }
+    }
+
+    /**
+     * DescribedPredicate defining condition for a class to have immutable fields.
+     */
+    private static final class ImmutableFieldsPredicate extends DescribedPredicate<JavaClass> {
+        private ImmutableFieldsPredicate() {
+            super("immutable fields");
+        }
+
+        @Override
+        public boolean apply(JavaClass input) {
+            final Set<JavaField> fields = input.getFields();
+            return fields.stream()
+                .filter(javaField -> {
+                    return !ModulePropertyPredicate.isModuleProperty(javaField)
+                        && !SUPPRESSED_FIELDS_IN_MODULES.contains(javaField.getFullName());
+                })
+                .allMatch(javaField -> {
+                    final Set<JavaModifier> javaFieldModifiers = javaField.getModifiers();
+                    boolean result = javaFieldModifiers.contains(JavaModifier.FINAL)
+                        && ImmutableFieldArchCondition.isImmutable(javaField);
+                    if (!javaField.getName().contains("MSG")) {
+                        result = result && javaFieldModifiers.contains(JavaModifier.PRIVATE);
+                    }
+                    return result;
+                });
+        }
+    }
+
+    /**
+     * ArchCondition checking if a class is present in the suppression list.
+     */
+    private static final class JavaClassSuppressionArchCondition extends ArchCondition<JavaClass> {
+
+        private final Set<String> suppressedClasses;
+
+        private JavaClassSuppressionArchCondition(Set<String> suppressedClasses,
+                                                  String description) {
+            super(description);
+            this.suppressedClasses = suppressedClasses;
+        }
+
+        @Override
+        public void check(JavaClass item, ConditionEvents events) {
+            if (!suppressedClasses.contains(item.getFullName())) {
+                final String message = String.format(
+                    Locale.ROOT, "should %s or resolved.", getDescription());
+                events.add(SimpleConditionEvent.violated(item, message));
+            }
+        }
+    }
```

Comment:
Please do this to avoid duplicated classes:
```suggestion

    private static final class ArchRuleSuppressionCondition<T extends HasName.AndFullName>
            extends ArchCondition<T> {

        private final Set<String> suppressions;

        private ArchRuleSuppressionCondition(Set<String> suppressions,
                                             String description) {
            super(description);
            this.suppressions = suppressions;
        }

        @Override
        public void check(T item, ConditionEvents events) {
            if (!suppressions.contains(item.getFullName())) {
                final String message = String.format(
                    Locale.ROOT, "should %s or resolved.", getDescription());
                events.add(SimpleConditionEvent.violated(item, message));
            }
        }
    }

```

Reply:
I already tried this, AndFullName is package private.

Reply:
> AndFullName is package private

No, this is not correct: https://github.com/TNG/ArchUnit/blob/c081b98f139546962c98c89f135f78edab27de49/archunit/src/main/java/com/tngtech/archunit/core/domain/properties/HasName.java#L33

Please try again.

Reply:
Thanks for clarifying :), done.